pam/internal/adapter/nativemodel.go

```diff
@@ -0,0 +1,633 @@
+		})
+	}
+
+	hasWait := m.uiLayout.GetWait() == "true"
+
+	switch m.uiLayout.Type {
+	case "form":
+		return m.handleFormChallenge(hasWait)
+
+	case "qrcode":
+		if !hasWait {
+			return sendEvent(pamError{
+				status: pam.ErrSystem,
+				msg:    "Can't handle qrcode without waiting",
+			})
+		}
+		return m.handleQrCode()
+
+	case "newpassword":
+		return m.handleNewPassword()
+
+	default:
+		return sendEvent(pamError{
+			status: pam.ErrSystem,
+			msg:    fmt.Sprintf("Unknown layout type: %q", m.uiLayout.Type),
+		})
+	}
+}
+
+func (m nativeModel) handleFormChallenge(hasWait bool) tea.Cmd {
+	if buttonLabel := m.uiLayout.GetButton(); buttonLabel != "" {
+		authMode := "selected authentication mode"
+		authModeIdx := slices.IndexFunc(m.authModes, func(mode *authd.GAMResponse_AuthenticationMode) bool {
+			return mode.Id == m.selectedAuthMode
+		})
+		if authModeIdx > -1 {
+			authMode = m.authModes[authModeIdx].Label
+		}
+		choices := choicesList{
+			{id: "continue", label: fmt.Sprintf("Proceed with %s", authMode)},
+		}
+		if buttonLabel := m.uiLayout.GetButton(); buttonLabel != "" {
+			choices = append(choices, choicePair{id: "button", label: buttonLabel})
+		}
+
+		id, err := m.promptForChoice(authMode, choices, "Select action")
+		if errors.Is(err, errGoBack) {
+			return m.goBackCommand()
+		}
+		if errors.Is(err, errEmptyResponse) {
+			return sendEvent(nativeChallengeRequested{})
+		}
+		if err != nil {
+			return sendPamError(err)
+		}
+		if id == "button" {
+			return sendEvent(reselectAuthMode{})
+		}
+	}
+
+	var prompt string
+	if m.uiLayout.Label != nil {
+		prompt, _ = strings.CutSuffix(*m.uiLayout.Label, ":")
+	}
+
+	if prompt == "" {
+		switch m.uiLayout.GetEntry() {
+		case "digits":
+			fallthrough
+		case "digits_password":
+			prompt = "PIN"
+		case "chars":
+			prompt = "Value"
+		case "chars_password":
+			prompt = "Password"
+		}
+	}
+
+	instructions := "Leave the input field empty or insert '%[1]s' to cancel the request and go back"
+	if hasWait {
+		// Duplicating some contents here, as it will be better for translators once we've them
+		instructions = "Leave the input field empty to wait for other authentication method" +
+			"or insert '%[1]s' to go back"
+		if m.uiLayout.GetEntry() == "" {
+			instructions = "Leave the input field empty to wait for the authentication method" +
+				"or insert '%[1]s' to go back"
+		}
+	}
+
+	err := m.sendInfo(instructions, nativeCancelKey)
+	if err != nil {
+		return sendPamError(err)
+	}
+
+	challenge, err := m.promptForChallenge(prompt)
+	if errors.Is(err, errGoBack) {
+		return m.goBackCommand()
+	}
+	if errors.Is(err, errEmptyResponse) {
+		if hasWait {
+			return sendAuthWaitCommand()
+		}
+		err = nil
+	}
+	if err != nil {
+		return sendPamError(err)
+	}
+
+	return sendEvent(isAuthenticatedRequested{
+		item: &authd.IARequest_AuthenticationData_Challenge{Challenge: challenge},
+	})
+}
+
+// TestNativeAuthenticate/Authenticate_user_switching_to_local_broker
```

Comment:
(?)

Reply:
Oh, it was a note on a test I wanted to retest :-D